### PR TITLE
Update the import table spec

### PR DESCRIPTION
**Pull request by Dhruvi16**
This will address #52

### CODE REVIEW

**Comment by Dhruvi16** [comment]
Before reviewing I have some questions to discuss -

1. What will be the format errors specifically? We discussed about some date errors with Brent, I am not so sure about them @ghislaineguerin
2. What will be the errors related to the data paste and URI paste while inputting?
3. Will there be any difference in validation of change in data of the table at the confirmation stage?
4.  Are we allowing user to add column or a record at confirmation stage? If not are we gonna do that in future?
5. Will the navigation link of data import be visible when the user is inside the table view? 
6. For duplicate columns, the UX in my mind would be not allowing user to select them before they change the name of the header. Is that an appropriate take?

**Comment by ghislaineguerin** [comment]
> 1. What will be the format errors specifically? We discussed about some date errors with Brent, I am not so sure about them @ghislaineguerin

- Improperly formatted could be due to mismatched number of headers and columns count

> 2. What will be the errors related to the data paste and URI paste while inputting?

- The URI points to a missing resource / broken link
- The URI text is in the correct form to potentially be a legitimate URI

> 3. Will there be any difference in validation of change in data of the table at the confirmation stage?

- Not at this point

> 4. Are we allowing user to add column or a record at confirmation stage? If not are we gonna do that in future?

- Not during the import step. We won't do this in the future.

> 5. Will the navigation link of data import be visible when the user is inside the table view?

- The user should always see their current location. We should ensure consistency with the latest navigation updates.

> 6. For duplicate columns, the UX in my mind would be not allowing user to select them before they change the name of the header. Is that an appropriate take?

- All columns are pre-selected by default. I'm assuming we'd show warnings in the duplicates and prevent user from completing the import process unless changed.

**Comment by ghislaineguerin** [comment]
From our call: 

- Update file upload UI
- Update file upload UI with the warnings in context
- Check if abandoned import process can also delete the created table
- Add scenario where table confirmation was force-closed
- In confirmation step, ensure users are aware of selected columns and warnings that are hidden by the scroll
- Remove option to rename table from confirmation step
- Add note dropdown vs. table inspector for data type options

File: design/specs/table-import.md
Comment (ghislaineguerin) [inline design/specs/table-import.md:14]: I think we can remove this part as we are adding type detection now and will handle other tasks separatedly.

File: design/specs/table-import.md
Comment (ghislaineguerin) [inline design/specs/table-import.md:16]: We can only have one h1 heading in the document. Please update.

File: design/specs/table-import.md
Comment (ghislaineguerin) [inline design/specs/table-import.md:22]: It would be enough to mention that user can access data import from the Schema page, since there will be a separate spec to address the design and layout for this page.

File: design/specs/table-import.md
Comment (ghislaineguerin) [inline design/specs/table-import.md:30]: I have discussed this with Pavish and the current backend won't handle renaming the table before it is created. We will have to move table rename to the confirmation step instead.

**Review by ghislaineguerin — CHANGES_REQUESTED** [review]
Thanks Dhruvi, looks good. I've added my feedback.

File: design/specs/table-import.md
Comment (kgodey) [inline design/specs/table-import.md:144]: This was the main problem that this spec needed to solve, we need more detail on the data type changing menu. 

Also, we're trying to have visual consistency with the table here, but I don't see anything about that in the spec.

File: design/specs/table-import.md
Comment (kgodey) [inline design/specs/table-import.md:146]: What does this mean? How does it depend on it?

File: design/specs/table-import.md
Comment (kgodey) [inline design/specs/table-import.md:148]: Please provide more detailed specs and wireframes for this.

File: design/specs/table-import.md
Comment (kgodey) [inline design/specs/table-import.md:159]: I'm not sure if I would call this reversion, it's just setting the data type.

File: design/specs/table-import.md
Comment (kgodey) [reply design/specs/table-import.md:148]: What about display options?

File: design/specs/table-import.md
Comment (kgodey) [inline design/specs/table-import.md:173]: How will it be highlighted? Also, this is inconsistent with other parts of the spec about the table getting deleted if not confirmed (but this one is correct).

File: design/specs/table-import.md
Comment (kgodey) [inline design/specs/table-import.md:181]: Please see note about vertical tabs above – they don't make sense for a mutually exclusive choice.

**Comment by Dhruvi16** [comment]
> This spec needs a lot of work. There's a lot of things here that can't be implemented quickly and don't make any sense for the demo use case. There's very little detail on the actual problem that we need design work on – visual consistency with tables and the interface for changing data types and display options.

Alright. For the data types and the display options, I had proposed a similar data type selector as we have now in the staging since it is consistent but @ghislaineguerin pointed out that we were ripping it off so there is pending discussion yet to happen on it. So, @ghislaineguerin suggested to put a note on it and we can decide it when we have table inspector.

For visual consistency, I am not sure about it. Will have to look into it.

File: design/specs/table-import.md
Comment (Dhruvi16) [reply design/specs/table-import.md:36]: I think I disagree here. Tabs are also used for navigation. Plus here the user is accomplishing the same task but has different choices. And tabs are one of the most scannable components out there. So I am not sure how users are gonna miss them.

The current UI that is the radio buttons are bit more confusing since its a type of an input and not something user would perceive as a navigating tool.

File: design/specs/table-import.md
Comment (Dhruvi16) [reply design/specs/table-import.md:56]: Okay I agree with this. This would put more cognitive load seeing our user persona. And we do not need this on the first screen since user is handling import here so we should let them focus on that rather than worrying about future errors caused by data inference.

File: design/specs/table-import.md
Comment (Dhruvi16) [reply design/specs/table-import.md:59]: In one of the discussions with @mathemancer and @ghislaineguerin we did discuss the probability of import getting stuck if the data inference is taking too much time. Simply opting out might let the import happen right?

File: design/specs/table-import.md
Comment (Dhruvi16) [reply design/specs/table-import.md:63]: Yeah I agree with this too. I was too skeptical about this. I believe we do not need this. Again the same comment as data inference.

File: design/specs/table-import.md
Comment (Dhruvi16) [reply design/specs/table-import.md:67]: As per a discussion with @ghislaineguerin, we discussed that sometimes there might be JSON data which might confuse the algorithm and hence the file would not be imported. So that is why I thought this might be a better way to handle it.

File: design/specs/table-import.md
Comment (Dhruvi16) [reply design/specs/table-import.md:81]: Oh okay. But I think it is good to have since we have everything here regarding the import.

File: design/specs/table-import.md
Comment (Dhruvi16) [reply design/specs/table-import.md:115]: Alright. Got it.

File: design/specs/table-import.md
Comment (Dhruvi16) [reply design/specs/table-import.md:117]: Yep. I agree. Will make those changes.

File: design/specs/table-import.md
Comment (kgodey) [reply design/specs/table-import.md:36]: > Plus here the user is accomplishing the same task but has different choices. 

This is my point exactly, the user has to make a single choice of import method here and we need to make it clear that they have three choices and they can pick one. I don't think we need to use the radio buttons (although they do fulfill that purpose) but we need something that shows all of the options in a visually prominent way. With vertical tabs, it's easy for people to focus on the center of the page and miss that there are other options on the side.

File: design/specs/table-import.md
Comment (kgodey) [reply design/specs/table-import.md:67]: This is probably going to confuse our user persona (same as my comments above). It seems like showing an error for un-importable files would work fine.

**Comment by kgodey** [comment]
@Dhruvi16 Please also take a look at https://github.com/centerofci/mathesar/issues/934, this was the original issue that described the problems the new design needed to solve. 

Also, the table inspector spec https://github.com/centerofci/mathesar-wiki/pull/51 and the inline editing spec https://github.com/centerofci/mathesar-wiki/pull/55 are pretty much done, even though they're not merged. You should be able to reference them enough to make the decisions you need.

File: design/specs/table-import.md
Comment (Dhruvi16) [reply design/specs/table-import.md:30]: I did not get this, could we discuss this please?

File: design/specs/table-import.md
Comment (Dhruvi16) [reply design/specs/table-import.md:36]: temp
I am not sure how vertical tabs are not visible enough. This is one example of tabs used for import file itself. We can I guess play around with the size of tabs and move the tabs inside the rectangle so they are visible enough?

Comment (kgodey) [reply design/specs/table-import.md:36]: Those are more visible and a design like that would resolve my concerns about visibility. It still comes across as a navigation pattern, and not a pattern related to the the user needing to pick a single option out of a few available choices.

**Comment by Dhruvi16** [comment]
So I think I need an approval on UI of the upload first so I can make other changes. I am figuring out data type change at the confirmation stage parallelly. But yeah need to discuss the UI here. 

Option 1 -


So this is again vertical tabs that I have seen widely across all the platforms that provides options for importing file. I have made tabs very prominent so they are visible and included them in the square. 

Option 2 -


Here I have made three options as action cards where the user will click on one of them and followed by it will be a modal where they upload the file or paste URI or Paste data. The progress and errors will be in the same modal. 

One very big advantage of this is that while progress the user wont be able to click anywhere. And as Kriti wanted the journeys will be mutually exclusive. 

@kgodey @ghislaineguerin @pavish please lemme know if you have any comments

**Comment by kgodey** [comment]
@Dhruvi16 Before I make up my mind, I have a question. Why does Option 2 need a modal? It seems like you could replace the three action cards with the file upload interface since the cards are no longer relevant.

**Comment by Dhruvi16** [comment]
> @Dhruvi16 Before I make up my mind, I have a question. Why does Option 2 need a modal? It seems like you could replace the three action cards with the file upload interface since the cards are no longer relevant.

You mean rather than cards, I use Drag & drop, URI paste and Data paste inputs right? That was my initial idea but I do feel that would be very cluttered since visually the inputs are not consistent plus for URI and data paste we need a continue button when the user pastes something. 
Modal gives a very clear and independent journey for each upload method I felt that is why I proposed modal rather that putting all the inputs on the screen.

**Comment by kgodey** [comment]
@Dhruvi16 This is what I mean:

**Comment by Dhruvi16** [comment]
> @Dhruvi16 This is what I mean: 
> 
> Once the user selects a method, then the action cards disappear entirely and is replaced by the input. You may want to provide a way of going back to selecting a different action method if needed.

Oh okay! I think this is a better idea. I think then we can create three steps rather than 2. Like Select upload method -> Upload data -> Confirm data.

So the user can go back forth the first two steps.

**Comment by kgodey** [comment]
> Oh okay! I think this is a better idea. I think then we can create three steps rather than 2. Like Select upload method -> Upload data -> Confirm data.

Yeah, that sounds good to me. I like this option the best because all the options are visually prominent and the other options go away once they've made their choice, which reduces cognitive load.

**Comment by Dhruvi16** [comment]
> > Oh okay! I think this is a better idea. I think then we can create three steps rather than 2. Like Select upload method -> Upload data -> Confirm data.
> 
> Yeah, that sounds good to me. I like this option the best because all the options are visually prominent and the other options go away once they've made their choice, which reduces cognitive load.

Alright I will update the spec with this UI then.

**Comment by kgodey** [comment]
> Alright I will update the spec with this UI then.

@ghislaineguerin probably also has an opinion. I think it's fine to update the spec if it unblocks you for now, as long as you're still expecting feedback.

**Comment by Dhruvi16** [comment]
@ghislaineguerin I have updated the spec. Please take a look. Also Could you please tell me where can I see UI for schema page? I need to update the spec with unconfirmed table part.

**Comment by ghislaineguerin** [comment]
I think we should consider how the user will interact with the data import screen.
After their first interaction with Data Import, it is likely that the user will be familiar with the different upload methods, or even have a preference for one of them. I worry about making all the methods accesible exclusively in the second step.

### Proposed Flow

In the same view the user should be able to drag a file, or click on the browse button to select a file. They should also be able to use a URL or paste data.

The file type information should be global, and not specific to the upload method. At the top of the upload controls we should have the following:

- A title for the section "Upload the data file"
- The detailed list of supported file types (e.g. CSV, TSV, etc.)

The upload controls favor the drag & drop method over the others. I'm assuming that the user will prefer the drag & drop method most of the time. It's the only one that allows for multiple file upload once available, so it's a good bet for future improvements.

For the paste and URL options we can open a modal to provide an input. I know that usually we want to have all options happen in the same window, but even for file selection, the user will still deal with the systems file selection dialog. It is also a pattern we have in other parts of the application, for example when naming a new schema.

Here's the proposed UI for the upload controls:

image

**Comment by Dhruvi16** [comment]
@ghislaineguerin I do think the UI is a bit cluttered here. And there are inconsistent user journeys for each upload method which I think can be confusing. Lemme know what you think

**Comment by ghislaineguerin** [comment]
@Dhruvi16 I get what you are saying.
Our present design includes radio buttons for each upload method. Changing the radio button would bring up the relevant input control, which is a common way to display these. This is similar to your first proposed idea with the tabs.
Our thinking with having the different cards was to give access to all controls so that the user wouldn't have to navigate to each, but having an additional step seems harder to implement and doesn't solve the original problem.

I'm ok with going back to radio buttons as the original design if we find that the updates add unnecessary complexity.

**Comment by ghislaineguerin** [comment]
If we were to consider a solution that doesn't involve radio buttons I think https://uppy.io/ does it well.
- It combines the drop area with the other upload options
- It provides a way to replace all content with the desired upload control

**Comment by Dhruvi16** [comment]
If we are discussing to retain radio buttons, I would like to point some thoughts -
1. Radio buttons are preferably not used for changing views. [Radio buttons](https://material.io/components/radio-buttons)
2. Radio buttons are majorly used as form components. 
3. Tabs have always meant switching between different views. You can see the [design guidelines](https://material.io/components/tabs)

Our requirements -
1. Parallel navigation between views.
2. Only one view is visible at a time so choices are mutually exclusive.
3. The choices should be clear and prominent.

**Comment by Dhruvi16** [comment]
But if we feel we should stick to radio buttons I will change the spec UI to radio buttons. But I personally feel that we should once consider Tabs. @ghislaineguerin would like to know your thoughts. Attaching platforms that use it -
temp
image

**Comment by ghislaineguerin** [comment]
@kgodey @Dhruvi16 
I agree with keeping the radio buttons, just wanted to clarify where the requirement came from.
Originally, we discussed allowing users to create their first table directly from an empty schema page (no tables), thus we needed a way to present the different upload methods and discussed the cards layout (a card for each upload method with a description). However, since we opted to have a dedicated Data Import view, this requirement seems unnecessary.

We can consider upgrading the radio buttons once we have more flows that require uploading of files.

**Comment by pavish** [comment]
@Dhruvi16 

> Radio buttons are majorly used as form components.

The import method section is a form input. We are asking for the user's input on which method they would like to import data into Mathesar.

Based on the chosen input, we are displaying more form content. We are not switching views.

> Tabs have always meant switching between different views

Yes, but tabs are not meant for making a choice, which is what we require. They are only meant for navigating across independent entities.

I do not see the problem that tabs solve over the radio buttons. If we are talking about making the options more prominent, we may want to focus on that angle instead. 

@ghislaineguerin 

> Our thinking with having the different cards was to give access to all controls so that the user wouldn't have to navigate to each, but having an additional step seems harder to implement and doesn't solve the original problem.

The user will still have to perform an additional step even if we display all of the options together. For eg., in your [option 1](https://github.com/centerofci/mathesar-wiki/pull/53#issuecomment-1229842127), the user will still have to click on their preferred method before we show the related content. I believe option 2 would result in a cognitive overload for the user.

I still do not understand why the original problem that's mentioned here is a problem.

I believe the only issue we have in this stage of import is providing more context to the user helping them choose an option.

**Comment by Dhruvi16** [comment]
I agree with the facts @ghislaineguerin & @pavish stated. So should I go ahead and change the UI again to radio buttons?

**Comment by ghislaineguerin** [comment]
@Dhruvi16 yes, please. Also suggest anything that might help provide more context to users as @pavish mentioned.

File: design/specs/table-import.md
Comment (mathemancer) [reply design/specs/table-import.md]: Sorry, I'm pretty late to this convo; just catching up. It's possible to choose the incorrect delimiter and still have a valid import set up.

Imagine (if you will) a headerless TSV whose correct delimiter is "tab". Now, further suppose there is a JSON column in this TSV, with a fixed schema. In particular, it has the same number of commas on each row. Our algo will not be able to choose properly which delimiter to use, and in fact, will chose comma in that case. We could change that specific behavior (i.e., put tab above comma in the choice hierarchy), but I don't think that solves the problem in general.

I think we _should_ display and give the option to modify the delimiter which has been chosen by the algorithm, but we shouldn't assume it's a comma unless that's what's returned by the algo. I think people who use spreadsheets regularly should be okay with that, since it's a common step for importing data into spreadsheet software.

@kgodey Are you arguing that they shouldn't have a chance to change the delimiter, or am I misunderstanding?

Comment (mathemancer) [reply design/specs/table-import.md]: Okay, now I've taken a moment to read through the document and see this in context. I agree with @kgodey that the user shouldn't have to make this choice at the upload/initial stage. I think it's more appropriate to give the chance to modify at the confirmation stage if something has gone wrong.

**Comment by mathemancer** [comment]
> The file type information should be global, and not specific to the upload method. At the top of the upload controls we should have the following:
> 
> * A title for the section "Upload the data file"
> * The detailed list of supported file types (e.g. CSV, TSV, etc.)
> 
> The upload controls favor the drag & drop method over the others. I'm assuming that the user will prefer the drag & drop method most of the time. It's the only one that allows for multiple file upload once available, so it's a good bet for future improvements.

@ghislaineguerin 

A couple of technical points: I don't think we should filter files on suffix. While we can show some info about what we support, it might confuse users. In the wild, there are many (many) TSVs with a `.txt` suffix or no suffix at all, and we should make sure the user understands that this isn't relevant for how the file is processed. All import processing is based on the actual data in the file, not its filetype suffix or other metadata.

Maybe some users prefer drag and drop, but I definitely prefer the file browser. Also, I'm pretty used to being able to open (or import in our case) multiple files via a file browser (with shift-click or ctrl-click for a range or individual selections, respectively).

**Comment by mathemancer** [comment]
@Dhruvi16 Have the comments about the delimiter overriding been addressed? It seems like the spec is still implying that the user needs to deal with this before importing.

File: design/specs/table-import.md
Comment (ghislaineguerin) [inline design/specs/table-import.md:107]: It is not clear to me what 'View Table' will do. Will this force confirmation? If so, can we update the action to reflect that?

File: design/specs/table-import.md
Comment (ghislaineguerin) [inline design/specs/table-import.md:84]: What do you mean by no data options available?

File: design/specs/table-import.md
Comment (ghislaineguerin) [inline design/specs/table-import.md:41]: The drop area shouldn't be visible at this point right? (the dashed border)

**Review by ghislaineguerin — COMMENTED** [review]
Thanks @Dhruvi16 the updates look good. I've added some notes/questions.

**Comment by kgodey** [comment]
I will leave the next pass of review to @ghislaineguerin and @pavish. Please let me know once it looks good to both of you and I'll do a final pass from a product perspective.

**Review by pavish — APPROVED** [review]
I think the spec is good to merge. We will not be fully implementing this and we would probably need further updates when we are ready to implement it.

I've added a note in https://github.com/centerofci/mathesar-wiki/pull/53/commits/2e6ccb2a2250b3d947afb35f8e427eac0bdbe2b7 to archive the spec for now.

For the demo video & the live demo, the design deliverable would directly be the frontend PR which can be reviewed by @ghislaineguerin and @kgodey .

**Review by kgodey — APPROVED** [review]

**Comment by kgodey** [comment]
Approved on my end, since it still says "changes requested" by @ghislaineguerin, I'll leave it up to her to merge this spec.

**Review by ghislaineguerin — APPROVED** [review]

**Pull request merged**
